hivemind/client/averaging/allreduce.py

```diff
@@ -210,7 +205,20 @@ async def _send_error_to_peer(self, peer_endpoint: Endpoint, code: averaging_pb2
         await stream.done_writing()
 
     def finalize(self, *, cancel: bool = False, exception: Optional[BaseException] = None):
+        """ finish or terminate AllReduceRunner, propagate any errors / cancellations to peers. """
         assert not cancel or not exception, "finalize accepts either exception or cancel, but not both"
+        pending_tasks = set()
+        if cancel or exception:
+            # propagate error to peers
+            if cancel or isinstance(exception, asyncio.CancelledError):
+                code = averaging_pb2.CANCELLED
+            else:
+                code = averaging_pb2.INTERNAL_ERROR
+            logger.debug(f"{self} - notifying peers about {averaging_pb2.MessageCode.Name(code)}")
+            for peer_endpoint, mode in zip(self.ordered_group_endpoints, self.modes):
+                if peer_endpoint != self.endpoint and mode != AveragingMode.CLIENT:
+                    pending_tasks.add(asyncio.create_task(self._send_error_to_peer(peer_endpoint, code)))
```

Comment:
nit: Consider covering this with tests if it is really not covered now (as Codecov says).

Reply:
This can only be tested by catching averager midway through transfering tensors. I hope that nirvana can help us handle this.